Zend/zend_weakrefs.c

```diff
@@ -187,11 +187,20 @@ ZEND_API zend_result zend_weakrefs_hash_del(HashTable *ht, zend_object *key) {
 	return FAILURE;
 }
 
-ZEND_API void zend_weakrefs_hash_clean(HashTable *ht) {
+static void zend_weakrefs_hash_clean_ex(HashTable *ht, int type) {
 	zend_ulong obj_key;
-	ZEND_HASH_FOREACH_NUM_KEY(ht, obj_key) {
-		zend_weakrefs_hash_del(ht, zend_weakref_key_to_object(obj_key));
+	ZEND_HASH_MAP_FOREACH_NUM_KEY(ht, obj_key) {
+		/* Optimization: Don't call zend_weakref_unref_single to free individual entries from ht when unregistering (which would do a hash table lookup, call zend_hash_index_del, and skip over any bucket collisions).
+		 * Let freeing the corresponding values for WeakMap entries be done in zend_hash_clean, freeing objects sequentially.
+		 * The performance difference is notable for larger WeakMaps with worse cache locality. */
+		zend_weakref_unregister(
+			zend_weakref_key_to_object(obj_key), ZEND_WEAKREF_ENCODE(ht, type), 0);
 	} ZEND_HASH_FOREACH_END();
+	zend_hash_clean(ht);
```

Comment:
This is too much? Wasn't there in the first place, neither in zend_weakrefs_hash_clean, nor in zend_weakmap_free_obj.